### PR TITLE
Template cluster resource

### PR DESCRIPTION
**Pull request by leosunmo**
This PR will leverage Terraform's template engine to create the cluster resource file and apply it instead of creating it using command line flags.

This enables us to specify custom options for the API server among other things that are not currently available to us from the command line. It should also make it very easy to add any other flag you might want from the list of available flags: https://github.com/kubernetes/kops/blob/master/docs/cluster_spec.md#kubeapiserver
For a more complete list of available flags, see the source code:
https://github.com/kubernetes/kops/blob/612f8b1952d6cbe0fb4974cfc709fbb5441d7227/pkg/apis/kops/componentconfig.go#L178

Right now I generate a `values.yaml` file using Terraforms templating language and then go on to use the `values.yaml` as variables in Kops templating engine to generate the final cluster spec file. Any improvements to this mess would be greatly appreciated. 

I've also run `terraform fmt` on all files to tidy them up.

### CODE REVIEW

**Comment by leosunmo** [comment]
I wasn't testing this properly so here's some additions to make it work if running from scratch (I had leftover config that made it work in testing). It will add `kops create secret` command to make `kops update` work as it needs all the that `kops create cluster` creates in the S3 bucket. This should now produce the same results as if running `kops create cluster` but with a customised/template-able Cluster yaml configuration.

**Comment by FutureSharks** [comment]
Hey @leosunmo ! 
I've pushed fmt changes to master. If you rebase it will make this PR much easier to read.

File: example/kops_clusters.tf [hunk 1]
```diff
@@ -21,10 +21,13 @@ module "cluster1" {
   internet_gateway_id       = "${aws_internet_gateway.public.id}"
   public_subnet_cidr_blocks = ["${local.cluster1_public_subnet_cidr_blocks}"]
   kops_dns_mode             = "private"
+  auditlog_logpath          = "/var/log/kube-apiserver-audit.log"
+  oidc_clientid             = "kubernetes"
+  oidc_issuerurl            = "https://accounts.google.com"
```
Comment (FutureSharks) [inline example/kops_clusters.tf]: The 2 examples in here should be as simple as possible to show how to use the module. Therefore I think we shouldn't be include OIDC stuff here by default. Could you just document how it works?

File: module/cluster_config.yaml.tmpl [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+{{- range $index, $zone := splitList "," .zones }}
+apiVersion: kops/v1alpha2
+kind: InstanceGroup
+metadata:
+  labels:
+    kops.k8s.io/cluster: {{$.cluster_fqdn}}
+  name:  master-{{$zone}}
+spec:
+  image: kope.io/{{$.ami_name}}
```
Comment (FutureSharks) [inline module/cluster_config.yaml.tmpl:70]: What is this `kope.io`? Is it from the kops project?

File: module/kops.tf [hunk 3]
```diff
@@ -4,19 +4,58 @@ resource "null_resource" "check_kops_version" {
...
+  }
+}
+
 resource "null_resource" "create_cluster" {
-  depends_on = [ "null_resource.check_kops_version" ]
+  depends_on = ["null_resource.check_kops_version", "null_resource.generate_template"]
+
   provisioner "local-exec" {
-    command = "kops create cluster --cloud=aws --dns ${var.kops_dns_mode} --authorization RBAC --networking ${var.kubernetes_networking} --zones=${join(",", data.aws_availability_zones.available.names)} --node-count=${var.node_asg_desired} --master-zones=${local.master_azs} --target=terraform --api-loadbalancer-type=public --vpc=${var.vpc_id} --state=s3://${var.kops_s3_bucket_id} --kubernetes-version ${var.kubernetes_version} ${local.cluster_fqdn}"
+    command = "kops create -f cluster_config.yaml --state=s3://${var.kops_s3_bucket_id} && kops create secret --state=s3://vend-test-kops-sydney --name cluster.ap-southeast-2.test.vendhq.com sshpublickey admin -i ~/.ssh/id_rsa.pub && kops update cluster ${var.cluster_fqdn} --state=s3://${var.kops_s3_bucket_id} --target=terraform --yes"
```
Comment (FutureSharks) [inline module/kops.tf]: Will this make every cluster use the templates, regardless of whether those features are wanted?

Comment (FutureSharks) [reply module/kops.tf]: Also here you have a hard coded s3 bucket name `vend-test-kops-sydney`

Comment (FutureSharks) [reply module/kops.tf]: Is the secret required for every cluster? Even if we want just a vanilla kops cluster without these extra features? If not, then I would move this command to separate `null_resource`.

**Review by FutureSharks — CHANGES_REQUESTED** [review]

**Comment by FutureSharks** [comment]
I am keen to support some extra features but my number 1 goal with this module is stability and hence I want the default behaviour to be as close to the kops default configuration as possible (or exactly the kops default). Do you think you could make the features offered in this PR optional instead of mandatory?

File: module/cluster_config.yaml.tmpl (hunk 2 above)
Comment (leosunmo) [reply module/cluster_config.yaml.tmpl:70]: This one threw me off for the longest time, but yes, that's the URL they use.
https://github.com/kubernetes/kops/blob/2a5e13660d59184f6269c406a95ff49181408c8d/channels/stable#L19

**Comment by leosunmo** [comment]
I agree with the goal of keeping it simple and stable. The changes I've made keeps incredibly close to the Kops defaults except for additional APIserver flags. You can easily compare the outputs by running the current `kops create cluster` command and then generating the templates from my branch using something like `terraform plan -target=module.kubernetes-cluster.null_resource.generate_template` and looking at the changes. It should match what's in your S3 bucket almost perfectly. 

Now, a lot of the config in S3 is never actually going to be used by Kops since we roll our own ASG etc. in this Terraform module, so they aren't as important. The important stuff is the `user_data` that we replicate here instead of letting Kops generate it in code. The `user_data` is already being dealt with in this module and I have not touched it at all so there should be very few high risk changes in this PR. The only risky part I can think of is if the templates aren't generated correctly and `kops update` doesn't generate the rest of the data needed in S3 for `protokube` to execute properly. So far I haven't come across anything strange in my testing as the templates are very reliable.

Regarding making these changes optional. I'd much prefer them to be optional, that way I could create a PR for my `user_data` changes that enabled Heptio's AWS Authenticator as well. Since Terraform is incredible limited when it comes to conditionals I haven't figured out a way to do that properly, especially when you want multiple user data templates to generate and add to the final output, but you want to be able to arbitrarily disable templates. I have yet to come up with a non "bash script for all the things" way of doing this. This is also true for the template additions in this PR.

File: module/kops.tf [hunk 4]
```diff
@@ -4,11 +4,47 @@ resource "null_resource" "check_kops_version" {
...
+  }
+}
+
 resource "null_resource" "create_cluster" {
-  depends_on = ["null_resource.check_kops_version"]
+  depends_on = ["null_resource.check_kops_version", "null_resource.generate_template"]
 
   provisioner "local-exec" {
-    command = "kops create cluster --cloud=aws --dns ${var.kops_dns_mode} --authorization RBAC --networking ${var.kubernetes_networking} --zones=${join(",", data.aws_availability_zones.available.names)} --node-count=${var.node_asg_desired} --master-zones=${local.master_azs} --target=terraform --api-loadbalancer-type=public --vpc=${var.vpc_id} --state=s3://${var.kops_s3_bucket_id} --kubernetes-version ${var.kubernetes_version} ${local.cluster_fqdn}"
+    command = "kops create -f cluster_config.yaml --state=s3://${var.kops_s3_bucket_id} && kops create secret --state=s3://vend-test-kops-sydney --name cluster.ap-southeast-2.test.vendhq.com sshpublickey admin -i ~/.ssh/id_rsa.pub && kops update cluster ${var.cluster_fqdn} --state=s3://${var.kops_s3_bucket_id} --target=terraform --yes"
```
Comment (FutureSharks) [inline module/kops.tf]: There is still odd hard coded values in here. e.g. `vend-test-kops-sydney` and `cluster.ap-southeast-2.test.vendhq.com`.

Also, you are still using the `cluster_config.yaml` template by default. This needs to be optional, if enabled.

**Review by FutureSharks — CHANGES_REQUESTED** [review]

File: module/kops.tf (hunk 4 above)
Comment (leosunmo) [reply module/kops.tf]: Ops! I must've copy pasted from my test setup and forgotten! Good catch!

Regarding the `cluster_config.yaml` template being optional, as I said above I don't think I can make it optional in a neat fashion as Terraform is limited when it comes to conditionals.
Is there anything in particular that is concerning regarding using the template file?

**Comment by FutureSharks** [comment]
I'm sorry Leo but I just feel this is too complicated. I like OIDC auth, and I [use it myself](https://github.com/deliveryhero/dhh-system-engineering/tree/master/kubernetes/examples/k8s-oidc-authenticator), but trying to do it in this module by adding 8 more variables, templates and more commands is too much when I don't think many people will use it. Also due to the limitations in TF it doesn't seem it's possible to do it in a nice clean way.

It also means that then I have to check the format and contents of the templates for every new kops version to make sure it's correct. Or understand better what parts of the templates are actually used or not.

**Comment by FutureSharks** [comment]
Could you move the template file completely out of the module and then just make it possible to pass it in as variable?

I made a quick example here: https://github.com/FutureSharks/tf-kops-cluster/compare/cluster_create_yaml

**Comment by leosunmo** [comment]
Hey, unfortunately passing the template as a variable isn't as simple as that since it requires subsequent commands to be different (`kops update` etc.). I could potentially build the whole `create` command like that, but at that point it'll be quite messy and _very_ unclear which options are supported (I can't customise OIDC or AWS Auth unless I pass a template file variable for example). In that case it's not worth using the template and the extra options at all. I can maintain a fork of this on my account which has a more opinionated approach which forces you to use templates and can work with AWS IAM as well.

**Comment by FutureSharks** [comment]
> I can maintain a fork of this on my account which has a more opinionated approach which forces you to use templates and can work with AWS IAM as well.

I think this is the best option for now 🙂

I was even thinking of making the module more simple by only supporting a single networking mode. The problem I have is that every time there is a new kops version, even a minor one, it's quite a lot of work to update this module with the changes. Lot's of things have changed over the versions of kops, e.g. SG rules, IAM policy, resource tagging, user-data and more. And every time I have to go through the terraform output of each kops version with each network mode and compare them all to see what's changed.

**Pull request closed without merging**